# 001-compile-prereqs: Compile prerequisites: atomic STPU fields, composite keys, call-generation API + co-located destructors
# Files: 26
# Generated by scripts/split-consolidated-patches.py
diff --git a/Source/Reasoner/Classifier/CClassificationComputationItem.cpp b/Source/Reasoner/Classifier/CClassificationComputationItem.cpp
index dd17c3ff..9a6c5a78 100644
--- a/Source/Reasoner/Classifier/CClassificationComputationItem.cpp
+++ b/Source/Reasoner/Classifier/CClassificationComputationItem.cpp
@@ -19,6 +19,7 @@
  */
 
 #include "CClassificationComputationItem.h"
+#include <atomic>
 
 
 namespace Konclude {
@@ -29,6 +30,8 @@ namespace Konclude {
 
 
 			CClassificationComputationItem::CClassificationComputationItem() {
+				static std::atomic<cint64> sNextGeneration(0);
+				mGeneration = ++sNextGeneration;
 			}
 
 
diff --git a/Source/Reasoner/Classifier/CClassificationComputationItem.h b/Source/Reasoner/Classifier/CClassificationComputationItem.h
index 99509099..c119dfa6 100644
--- a/Source/Reasoner/Classifier/CClassificationComputationItem.h
+++ b/Source/Reasoner/Classifier/CClassificationComputationItem.h
@@ -22,7 +22,7 @@
 #define KONCLUDE_REASONER_CLASSIFIER_CCLASSIFICATIONCOMPUTATIONITEM_H
 
 // Libraries includes
-
+#include <atomic>
 
 // Namespace includes
 #include "ClassifierSettings.h"
@@ -60,6 +60,8 @@ namespace Konclude {
 			class CClassificationComputationItem {
 				// public methods
 				public:
+					cint64 mGeneration;
+
 					//! Constructor
 					CClassificationComputationItem();
 
diff --git a/Source/Reasoner/Classifier/CClassificationManager.h b/Source/Reasoner/Classifier/CClassificationManager.h
index 1081232c..51706874 100644
--- a/Source/Reasoner/Classifier/CClassificationManager.h
+++ b/Source/Reasoner/Classifier/CClassificationManager.h
@@ -96,8 +96,8 @@ namespace Konclude {
 
 					CSubsumptionClassifierFactory* mClassifierFac;
 
-					QHash<CConcreteOntology*,CSubsumptionClassifier*> mOntoClassifierHash;
-					QHash<CConcreteOntology*,CSubsumptionClassifier*> mOntoBackgroundClassifierHash;
+					QHash<QPair<CConcreteOntology*,cint64>,CSubsumptionClassifier*> mOntoClassifierHash;
+					QHash<QPair<CConcreteOntology*,cint64>,CSubsumptionClassifier*> mOntoBackgroundClassifierHash;
 
 					QSet<CSubsumptionClassifier*> mOntoClassifierSet;
 
@@ -105,8 +105,8 @@ namespace Konclude {
 
 
 
-					QHash<CConcreteOntology*,CSubsumptionClassifier*> mOntoDataPropertyClassifierHash;
-					QHash<CConcreteOntology*,CSubsumptionClassifier*> mOntoObjectPropertyClassifierHash;
+					QHash<QPair<CConcreteOntology*,cint64>,CSubsumptionClassifier*> mOntoDataPropertyClassifierHash;
+					QHash<QPair<CConcreteOntology*,cint64>,CSubsumptionClassifier*> mOntoObjectPropertyClassifierHash;
 
 
 				// private methods
diff --git a/Source/Reasoner/Classifier/CIncrementalKPSetClassSubsumptionClassifierThread.cpp b/Source/Reasoner/Classifier/CIncrementalKPSetClassSubsumptionClassifierThread.cpp
index d7fecd80..fc4acc88 100644
--- a/Source/Reasoner/Classifier/CIncrementalKPSetClassSubsumptionClassifierThread.cpp
+++ b/Source/Reasoner/Classifier/CIncrementalKPSetClassSubsumptionClassifierThread.cpp
@@ -59,7 +59,7 @@ namespace Konclude {
 				ontClassItem->initTaxonomyConcepts(ontology,taxonomy);
 				ontItemList.append(ontClassItem);
 				processingOntItemList.append(ontClassItem);
-				ontItemHash.insert(ontology,ontClassItem);
+				ontItemHash.insert({ontology, ontology->getCallGeneration()},ontClassItem);
 
 				readCalculationConfig(ontClassItem->getCalculationConfiguration());
 
@@ -183,7 +183,7 @@ namespace Konclude {
 
 					CPartialPruningTaxonomy *parTax = dynamic_cast<CPartialPruningTaxonomy *>(taxonomy);
 					if (parTax) {
-						COntologyClassClassificationItem *ontClassItem = (COntologyClassClassificationItem*)ontItemHash.value(ontology);
+						COntologyClassClassificationItem *ontClassItem = (COntologyClassClassificationItem*)ontItemHash.value({ontology, ontology->getCallGeneration()});
 						parTax->createStatistics(ontClassItem->getClassifierStatistics());
 					}
 
diff --git a/Source/Reasoner/Classifier/COptimizedClassExtractedSaturationSubsumptionClassifierThread.cpp b/Source/Reasoner/Classifier/COptimizedClassExtractedSaturationSubsumptionClassifierThread.cpp
index 98cfaa11..418e6d98 100644
--- a/Source/Reasoner/Classifier/COptimizedClassExtractedSaturationSubsumptionClassifierThread.cpp
+++ b/Source/Reasoner/Classifier/COptimizedClassExtractedSaturationSubsumptionClassifierThread.cpp
@@ -63,7 +63,7 @@ namespace Konclude {
 				ontClassItem->initTaxonomyConcepts(ontology,taxonomy);
 				ontItemList.append(ontClassItem);
 				processingOntItemList.append(ontClassItem);
-				ontItemHash.insert(ontology,ontClassItem);
+				ontItemHash.insert({ontology, ontology->getCallGeneration()},ontClassItem);
 
 				CCalculationConfigurationExtension* calcConfig = ontClassItem->getCalculationConfiguration();
 
@@ -75,7 +75,7 @@ namespace Konclude {
 
 				CPartialPruningTaxonomy *parTax = dynamic_cast<CPartialPruningTaxonomy *>(taxonomy);
 				if (parTax) {
-					COntologyClassClassificationItem *ontClassItem = (COntologyClassClassificationItem*)ontItemHash.value(ontology);
+					COntologyClassClassificationItem *ontClassItem = (COntologyClassClassificationItem*)ontItemHash.value({ontology, ontology->getCallGeneration()});
 					parTax->createStatistics(ontClassItem->getClassifierStatistics());
 				}
 
@@ -258,6 +258,8 @@ namespace Konclude {
 						if (optSubClassItem || optSubClassItem->isTaxonomyConstructionFailed()) {
 							finishOntologyClassification(optSubClassItem);
 							processingOntItemList.removeFirst();
+							ontItemHash.remove({optSubClassItem->getOntology(), optSubClassItem->getOntology()->getCallGeneration()});
+							mStaleClassItems.push_back(optSubClassItem);
 						} else {
 							processingOntItemList.removeFirst();
 							processingOntItemList.append(ontClassItem);
diff --git a/Source/Reasoner/Classifier/COptimizedKPSetClassSubsumptionClassifierThread.cpp b/Source/Reasoner/Classifier/COptimizedKPSetClassSubsumptionClassifierThread.cpp
index b518e242..ce7bfc9f 100644
--- a/Source/Reasoner/Classifier/COptimizedKPSetClassSubsumptionClassifierThread.cpp
+++ b/Source/Reasoner/Classifier/COptimizedKPSetClassSubsumptionClassifierThread.cpp
@@ -65,8 +65,8 @@ namespace Konclude {
 				mConfPossSubsumCalcOrderTopDown = true;
 				mConfPossSubsumCalcOrderBottomUp = false;
 
-				mConfPossSubsumPseudoModelPretest = true;
-				mConfPossSubsumPseudoModelTest = true;
+				mConfPossSubsumPseudoModelPretest = false;
+				mConfPossSubsumPseudoModelTest = false;
 				mConfWriteDebuggingData = false;
 
 				mStatProcesedSubsumMessCount = 0;
@@ -156,7 +156,7 @@ namespace Konclude {
 				ontClassItem->initTaxonomyConcepts(ontology,taxonomy);
 				ontItemList.append(ontClassItem);
 				processingOntItemList.append(ontClassItem);
-				ontItemHash.insert(ontology,ontClassItem);
+				ontItemHash.insert({ontology, ontology->getCallGeneration()},ontClassItem);
 
 				readCalculationConfig(ontClassItem->getCalculationConfiguration());
 
@@ -167,7 +167,7 @@ namespace Konclude {
 
 				CPartialPruningTaxonomy *parTax = dynamic_cast<CPartialPruningTaxonomy *>(taxonomy);
 				if (parTax) {
-					COntologyClassClassificationItem *ontClassItem = (COntologyClassClassificationItem*)ontItemHash.value(ontology);
+					COntologyClassClassificationItem *ontClassItem = (COntologyClassClassificationItem*)ontItemHash.value({ontology, ontology->getCallGeneration()});
 					parTax->createStatistics(ontClassItem->getClassifierStatistics());
 				}
 
@@ -753,8 +753,6 @@ namespace Konclude {
 
 
 			bool COptimizedKPSetClassSubsumptionClassifierThread::createNextSubsumtionTest() {
-
-
 				COntologyClassificationItem *loopOntClassItem = 0;
 				bool workTestCreated = false;
 				while (!workTestCreated && !processingOntItemList.isEmpty()) {
@@ -1174,6 +1172,8 @@ namespace Konclude {
 						if (optKPSetClassificationItem->hasPossibleSubsumptionTestingPhaseFinished() || optKPSetClassificationItem->isTaxonomyConstructionFailed()) {
 							finishOntologyClassification(optKPSetClassificationItem);
 							processingOntItemList.removeFirst();
+							ontItemHash.remove({optKPSetClassificationItem->getOntology(), optKPSetClassificationItem->getOntology()->getCallGeneration()});
+							mStaleClassItems.push_back(optKPSetClassificationItem);
 						} else {
 							processingOntItemList.removeFirst();
 							processingOntItemList.append(ontClassItem);
@@ -1714,7 +1714,7 @@ namespace Konclude {
 					statColl = ontClassItem->getCalculationStatisticsCollection();
 				}
 				job->setCalclulationStatisticsCollector(statColl);
-				CTestCalculatedCallbackEvent *testResultCallback = new CTestCalculatedCallbackEvent(this,job,workItem,statColl);
+				CTestCalculatedCallbackEvent *testResultCallback = new CTestCalculatedCallbackEvent(this,job,workItem,statColl,workItem->mGeneration);
 				job->setCalculationConfiguration(ontClassItem->getCalculationConfiguration());
 				CSubsumptionClassifierThread::processCalculationJob(job,ontClassItem,testResultCallback,false);
 				return this;
@@ -2415,10 +2415,12 @@ namespace Konclude {
 						}
 					}
 					succItem->decUnprocessedPredecessorItems();
-					if (succItem->hasOnlyProcessedPredecessorItems()) {
-						nextItemList->append(succItem);
-					} else {
-						nextCandItemSet->insert(succItem);
+					if (!succItem->isSatisfiableTestOrdered()) {
+						if (succItem->hasOnlyProcessedPredecessorItems()) {
+							nextItemList->append(succItem);
+						} else {
+							nextCandItemSet->insert(succItem);
+						}
 					}
 				}
 
@@ -2568,16 +2570,12 @@ namespace Konclude {
 					for (QList<COptimizedKPSetClassTestingItem*>::const_iterator it = itemList.constBegin(), itEnd = itemList.constEnd(); it != itEnd; ++it) {
 						COptimizedKPSetClassTestingItem* item = *it;
 
-						//QString iriClassNameString = CIRIName::getRecentIRIName(item->getTestingConcept()->getClassNameLinker());
-						//if (iriClassNameString == "http://usefulinc.com/ns/doap#Project") {
-						//	bool bug = true;
-						//}
-
 						if (!item->isEquivalentItem() && item != topItem) {
 							cint64 itemSubsumingCount = item->getSubsumingConceptItemCount();
 							CHierarchyNode* itemHierNode = item->getSatisfiableConceptHierarchyNode();
 
 							QList<COptimizedKPSetClassTestingItem*>* subsumingConceptItemList = item->getSubsumingConceptItemList();
+
 							QList<COptimizedKPSetClassTestingItem*>::const_iterator subsumedIt = subsumingConceptItemList->constBegin(), subsumedItEnd = subsumingConceptItemList->constEnd();
 
 							cint64 eqCount = 0;
diff --git a/Source/Reasoner/Classifier/COptimizedKPSetClassSubsumptionClassifierThread.h b/Source/Reasoner/Classifier/COptimizedKPSetClassSubsumptionClassifierThread.h
index e0e9782d..42b9efce 100644
--- a/Source/Reasoner/Classifier/COptimizedKPSetClassSubsumptionClassifierThread.h
+++ b/Source/Reasoner/Classifier/COptimizedKPSetClassSubsumptionClassifierThread.h
@@ -22,7 +22,7 @@
 #define KONCLUDE_REASONER_CLASSIFIER_COPTIMIZEDKPSETCLASSSUBSUMPTIONCLASSIFIERTHREAD_H
 
 // Libraries includes
-#include <QLinkedList>
+#include "QtCompat.h"
 
 // Namespace includes
 #include "ClassifierSettings.h"
diff --git a/Source/Reasoner/Classifier/COptimizedKPSetRoleOntologyClassificationItem.cpp b/Source/Reasoner/Classifier/COptimizedKPSetRoleOntologyClassificationItem.cpp
index 94ccc10f..0a87afe1 100644
--- a/Source/Reasoner/Classifier/COptimizedKPSetRoleOntologyClassificationItem.cpp
+++ b/Source/Reasoner/Classifier/COptimizedKPSetRoleOntologyClassificationItem.cpp
@@ -62,6 +62,8 @@ namespace Konclude {
 			}
 
 			COptimizedKPSetRoleOntologyClassificationItem::~COptimizedKPSetRoleOntologyClassificationItem() {
+				delete mTmpRoleClassificationOnt;
+				for (auto* item : mSatTestItemContainer) { delete item; }
 			}
 
 
diff --git a/Source/Reasoner/Classifier/COptimizedKPSetRoleSubsumptionClassifierThread.cpp b/Source/Reasoner/Classifier/COptimizedKPSetRoleSubsumptionClassifierThread.cpp
index b86992ec..48133731 100644
--- a/Source/Reasoner/Classifier/COptimizedKPSetRoleSubsumptionClassifierThread.cpp
+++ b/Source/Reasoner/Classifier/COptimizedKPSetRoleSubsumptionClassifierThread.cpp
@@ -355,7 +355,7 @@ namespace Konclude {
 				ontPropItem->initClassificationItem(ontology,hierarchy,mDataRolesClassification);
 				ontItemList.append(ontPropItem);
 				processingOntItemList.append(ontPropItem);
-				ontItemHash.insert(ontology,ontPropItem);
+				ontItemHash.insert({ontology, ontology->getCallGeneration()},ontPropItem);
 
 				readCalculationConfig(ontPropItem->getCalculationConfiguration());
 
@@ -904,6 +904,8 @@ namespace Konclude {
 						if (optKPSetClassificationItem->hasPossibleSubsumptionTestingPhaseFinished() || ontPropItem->isHierarchyConstructionFailed()) {
 							finishOntologyClassification(ontPropItem);
 							processingOntItemList.removeFirst();
+							ontItemHash.remove({ontPropItem->getOntology(), ontPropItem->getOntology()->getCallGeneration()});
+							mStaleClassItems.push_back(ontPropItem);
 						} else {
 							processingOntItemList.removeFirst();
 							processingOntItemList.append(ontPropItem);
@@ -1109,7 +1111,7 @@ namespace Konclude {
 					statColl = ontPropItem->getCalculationStatisticsCollection();
 				}
 				job->setCalclulationStatisticsCollector(statColl);
-				CTestCalculatedCallbackEvent *testResultCallback = new CTestCalculatedCallbackEvent(this,job,workItem,statColl);
+				CTestCalculatedCallbackEvent *testResultCallback = new CTestCalculatedCallbackEvent(this,job,workItem,statColl,workItem->mGeneration);
 				job->setCalculationConfiguration(ontPropItem->getCalculationConfiguration());
 				CSubsumptionClassifierThread::processCalculationJob(job,ontPropItem,testResultCallback,false);
 				return this;
diff --git a/Source/Reasoner/Classifier/COptimizedKPSetRoleSubsumptionClassifierThread.h b/Source/Reasoner/Classifier/COptimizedKPSetRoleSubsumptionClassifierThread.h
index c7a3c1d8..ea4be5d6 100644
--- a/Source/Reasoner/Classifier/COptimizedKPSetRoleSubsumptionClassifierThread.h
+++ b/Source/Reasoner/Classifier/COptimizedKPSetRoleSubsumptionClassifierThread.h
@@ -22,7 +22,7 @@
 #define KONCLUDE_REASONER_CLASSIFIER_COptimizedKPSetRoleSubsumptionClassifierThread_H
 
 // Libraries includes
-#include <QLinkedList>
+#include "QtCompat.h"
 
 // Namespace includes
 #include "ClassifierSettings.h"
diff --git a/Source/Reasoner/Classifier/COptimizedSubClassSubsumptionClassifierThread.cpp b/Source/Reasoner/Classifier/COptimizedSubClassSubsumptionClassifierThread.cpp
index 66cd4808..1238a68e 100644
--- a/Source/Reasoner/Classifier/COptimizedSubClassSubsumptionClassifierThread.cpp
+++ b/Source/Reasoner/Classifier/COptimizedSubClassSubsumptionClassifierThread.cpp
@@ -113,7 +113,7 @@ namespace Konclude {
 				ontClassItem->initTaxonomyConcepts(ontology,taxonomy);
 				ontItemList.append(ontClassItem);
 				processingOntItemList.append(ontClassItem);
-				ontItemHash.insert(ontology,ontClassItem);
+				ontItemHash.insert({ontology, ontology->getCallGeneration()},ontClassItem);
 
 				readCalculationConfig(ontClassItem->getCalculationConfiguration());
 
@@ -123,7 +123,7 @@ namespace Konclude {
 
 				CPartialPruningTaxonomy *parTax = dynamic_cast<CPartialPruningTaxonomy *>(taxonomy);
 				if (parTax) {
-					COntologyClassClassificationItem *ontClassItem = (COntologyClassClassificationItem*)ontItemHash.value(ontology);
+					COntologyClassClassificationItem *ontClassItem = (COntologyClassClassificationItem*)ontItemHash.value({ontology, ontology->getCallGeneration()});
 					parTax->createStatistics(ontClassItem->getClassifierStatistics());
 				}
 
@@ -146,10 +146,10 @@ namespace Konclude {
 			void COptimizedSubClassSubsumptionClassifierThread::createObviousSubsumptionSatisfiableTestingOrder(COptimizedSubClassOntologyClassificationItem* ontClassItem) {
 				CPrecomputation* precomputation = ontClassItem->getOntology()->getPrecomputation();
 				CSaturationData* saturationData = precomputation->getSaturationModelData();
-			
+
 				bool subsumerSaturationExtraction = CConfigDataReader::readConfigBoolean(ontClassItem->getCalculationConfiguration(),"Konclude.Calculation.Classification.SaturationSubsumerExtraction",true);
 
-				if (saturationData && subsumerSaturationExtraction) {
+				if (saturationData) {
 					createObviousSubsumptionSatisfiableTestingOrderFromSaturationData(ontClassItem);
 				} else {
 					createObviousSubsumptionSatisfiableTestingOrderFromBuildData(ontClassItem);
@@ -488,35 +488,27 @@ namespace Konclude {
 
 
 			bool COptimizedSubClassSubsumptionClassifierThread::createNextSubsumtionTest() {
-
-
 				COntologyClassificationItem *loopOntClassItem = 0;
 				bool workTestCreated = false;
 				while (!workTestCreated && !processingOntItemList.isEmpty()) {
 					COntologyClassificationItem *ontClassItem = processingOntItemList.first();
 
 					if (ontClassItem == loopOntClassItem) {
-						// don't run into infinite loop without doing something
 						break;
 					}
 
 					COptimizedSubClassOntologyClassificationItem *optSubClassItem = (COptimizedSubClassOntologyClassificationItem *)ontClassItem;
 
-
 					while (!workTestCreated && optSubClassItem->hasRemainingSatisfiableTests() && !optSubClassItem->isTaxonomyConstructionFailed()) {
-						// get next satisfiable test
-
 						QList<COptimizedSubClassSatisfiableTestingItem*>* nextItemList = optSubClassItem->getNextSatisfiableTestingItemList();
 						QSet<COptimizedSubClassSatisfiableTestingItem*>* nextCandItemSet = optSubClassItem->getNextCandidateSatisfiableTestingItemSet();
 						QSet<COptimizedSubClassSatisfiableTestingItem*>* remainingCandItemSet = optSubClassItem->getRemainingCandidateSatisfiableTestingItemSet();
 
-
 						COptimizedSubClassSatisfiableTestingItem* nextSatTestItem = nullptr;
 						if (!nextItemList->isEmpty()) {
 							nextSatTestItem = nextItemList->takeFirst();
 							remainingCandItemSet->remove(nextSatTestItem);
 							nextCandItemSet->remove(nextSatTestItem);
-
 						} else if (!nextCandItemSet->isEmpty()) {
 							COptimizedSubClassSatisfiableTestingItem* minItem = nullptr;
 							cint64 minUnpPredCount = 0;
@@ -532,26 +524,12 @@ namespace Konclude {
 							nextSatTestItem = minItem;
 							remainingCandItemSet->remove(nextSatTestItem);
 							nextCandItemSet->remove(nextSatTestItem);
-
 						} else if (!remainingCandItemSet->isEmpty()) {
-							COptimizedSubClassSatisfiableTestingItem* minItem = nullptr;
-							//cint64 minUnpPredCount = 0;
-							//for (QSet<COptimizedSubClassSatisfiableTestingItem*>::const_iterator it = remainingCandItemSet->constBegin(), itEnd = remainingCandItemSet->constEnd(); it != itEnd; ++it) {
-							//	COptimizedSubClassSatisfiableTestingItem* item = (*it);
-							//	cint64 predCount = item->getUnprocessedPredecessorItemCount();
-							//	if (!minItem || minUnpPredCount < predCount) {
-							//		minUnpPredCount = predCount;
-							//		minItem = item;
-							//	}
-							//}
-							//nextSatTestItem = minItem;
-							//remainingCandItemSet->remove(nextSatTestItem);
 							QSet<COptimizedSubClassSatisfiableTestingItem*>::const_iterator itNext = remainingCandItemSet->constBegin();
 							nextSatTestItem = *itNext;
-							remainingCandItemSet->remove(nextSatTestItem);							
+							remainingCandItemSet->remove(nextSatTestItem);
 						}
 
-
 						if (nextSatTestItem && !nextSatTestItem->isSatisfiableTestOrdered()) {
 							optSubClassItem->decRemainingSatisfiableTestsCount();
 							optSubClassItem->incRunningSatisfiableTestsCount();
@@ -560,14 +538,17 @@ namespace Konclude {
 							} else {
 								interpreteSatisfiableResult(optSubClassItem,nextSatTestItem->getSatisfiableConcept(),nextSatTestItem->getSatisfiableTestedResult());
 							}
+						} else if (!nextSatTestItem) {
+							break;
 						}
 					}
 
-
 					if (!workTestCreated) {
 						if (optSubClassItem->hasAllSatisfiableTestsCompleted() || optSubClassItem->isTaxonomyConstructionFailed()) {
 							finishOntologyClassification(optSubClassItem);
 							processingOntItemList.removeFirst();
+							ontItemHash.remove({optSubClassItem->getOntology(), optSubClassItem->getOntology()->getCallGeneration()});
+							mStaleClassItems.push_back(optSubClassItem);
 						} else {
 							processingOntItemList.removeFirst();
 							processingOntItemList.append(ontClassItem);
@@ -819,7 +800,7 @@ namespace Konclude {
 					statColl = ontClassItem->getCalculationStatisticsCollection();
 				}
 				job->setCalclulationStatisticsCollector(statColl);
-				CTestCalculatedCallbackEvent *testResultCallback = new CTestCalculatedCallbackEvent(this,job,workItem,statColl);
+				CTestCalculatedCallbackEvent *testResultCallback = new CTestCalculatedCallbackEvent(this,job,workItem,statColl,workItem->mGeneration);
 				job->setCalculationConfiguration(ontClassItem->getCalculationConfiguration());
 				CSubsumptionClassifierThread::processCalculationJob(job,ontClassItem,testResultCallback);
 				return this;
@@ -950,10 +931,12 @@ namespace Konclude {
 						succItem->setResultUnsatisfiableDerivated(true);
 					}
 					succItem->decUnprocessedPredecessorItems();
-					if (succItem->hasOnlyProcessedPredecessorItems()) {
-						nextItemList->append(succItem);
-					} else {
-						nextCandItemSet->insert(succItem);
+					if (!succItem->isSatisfiableTestOrdered()) {
+						if (succItem->hasOnlyProcessedPredecessorItems()) {
+							nextItemList->append(succItem);
+						} else {
+							nextCandItemSet->insert(succItem);
+						}
 					}
 				}
 
diff --git a/Source/Reasoner/Classifier/COptimizedSubClassSubsumptionClassifierThread.h b/Source/Reasoner/Classifier/COptimizedSubClassSubsumptionClassifierThread.h
index 25c88d21..81b79905 100644
--- a/Source/Reasoner/Classifier/COptimizedSubClassSubsumptionClassifierThread.h
+++ b/Source/Reasoner/Classifier/COptimizedSubClassSubsumptionClassifierThread.h
@@ -22,7 +22,7 @@
 #define KONCLUDE_REASONER_CLASSIFIER_COPTIMIZEDSUBCLASSSUBSUMPTIONCLASSIFIERTHREAD_H
 
 // Libraries includes
-#include <QLinkedList>
+#include "QtCompat.h"
 
 // Namespace includes
 #include "ClassifierSettings.h"
diff --git a/Source/Reasoner/Classifier/CSubsumptionClassifierThread.cpp b/Source/Reasoner/Classifier/CSubsumptionClassifierThread.cpp
index 7a665d9e..a7ae3118 100644
--- a/Source/Reasoner/Classifier/CSubsumptionClassifierThread.cpp
+++ b/Source/Reasoner/Classifier/CSubsumptionClassifierThread.cpp
@@ -90,7 +90,7 @@ namespace Konclude {
 
 
 			CSubsumptionClassifierThread *CSubsumptionClassifierThread::addOntologyClassifiedCallback(CConcreteOntology *ontology, CCallbackData *callback) {
-				COntologyClassificationItem *ontClassItem = ontItemHash.value(ontology,0);
+				COntologyClassificationItem *ontClassItem = ontItemHash.value({ontology, ontology->getCallGeneration()},0);
 				if (ontClassItem) {
 					ontClassItem->addClassifiedCallback(callback);
 				}
@@ -98,7 +98,7 @@ namespace Konclude {
 			}
 
 			CSubsumptionClassifierThread* CSubsumptionClassifierThread::addOntologyClassifyRequirements(CConcreteOntology *ontology, const QList<COntologyProcessingRequirement*>& requirementList) {
-				COntologyClassificationItem *ontClassItem = ontItemHash.value(ontology,nullptr);
+				COntologyClassificationItem *ontClassItem = ontItemHash.value({ontology, ontology->getCallGeneration()},nullptr);
 				if (ontClassItem) {
 					ontClassItem->addOntologyProcessingRequirements(requirementList);
 				}
@@ -106,24 +106,26 @@ namespace Konclude {
 			}
 
 			CSubsumptionClassifierThread *CSubsumptionClassifierThread::scheduleOntologyClassification(CConcreteOntology *ontology, CClassificationCalculationSupport *classificationSupport, CConfigurationBase *config) {
+				for (auto* stale : mStaleClassItems) { delete stale; }
+				mStaleClassItems.clear();
 				COntologyClassificationItem *ontClassItem = new COntologyClassificationItem(config,statistics);
 				ontItemList.append(ontClassItem);
 				processingOntItemList.append(ontClassItem);
-				ontItemHash.insert(ontology,ontClassItem);
+				ontItemHash.insert({ontology, ontology->getCallGeneration()},ontClassItem);
 				return this;
 			}
 
 
 
 			CSubsumptionClassifierThread* CSubsumptionClassifierThread::rescheduleOntologyClassification(CConcreteOntology *ontology) {
-				COntologyClassificationItem *ontClassItem = ontItemHash.value(ontology);
+				COntologyClassificationItem *ontClassItem = ontItemHash.value({ontology, ontology->getCallGeneration()});
 				processingOntItemList.append(ontClassItem);
 				return this;
 			}
 
 
 			bool CSubsumptionClassifierThread::isOntologyClassificationScheduled(CConcreteOntology *ontology) {
-				return ontItemHash.contains(ontology);
+				return ontItemHash.contains({ontology, ontology->getCallGeneration()});
 			}
 
 
@@ -214,7 +216,7 @@ namespace Konclude {
 				} else if (type == COntologyTellClassificationMessageEvent::EVENTTYPE) {
 					COntologyTellClassificationMessageEvent *otcme = (COntologyTellClassificationMessageEvent *)event;
 					CConcreteOntology *ontology = otcme->getOntology();
-					COntologyClassificationItem *ontClassItem = ontItemHash.value(ontology,0);
+					COntologyClassificationItem *ontClassItem = ontItemHash.value({ontology, ontology->getCallGeneration()},0);
 					if (ontClassItem) {
 						CClassificationMessageData* messageData = otcme->getMessageData();
 						CMemoryPool* memoryPools = otcme->getMemoryPool();
@@ -227,12 +229,18 @@ namespace Konclude {
 					CTestCalculatedCallbackEvent *tcue = (CTestCalculatedCallbackEvent *)event;
 					interpreteTestResults(tcue);
 					currRunningTestParallelCount--;
+#ifdef WASM_BUILD
+					if (currRunningTestParallelCount == 0) {
+						doNextPendingTests();
+					}
+#else
 					doNextPendingTests();
+#endif
 					return true;
 				} else if (type == CCallbackClassifiedOntologyEvent::EVENTTYPE) {
 					CCallbackClassifiedOntologyEvent *ccoe = (CCallbackClassifiedOntologyEvent *)event;
 					CCallbackData *possCallback = ccoe->getCallbackPossibleCallbackData();
-					COntologyClassificationItem *ontClassItem = ontItemHash.value(ccoe->getOntology(),0);
+					COntologyClassificationItem *ontClassItem = ontItemHash.value({ccoe->getOntology(), ccoe->getOntology()->getCallGeneration()},0);
 					if (possCallback) {
 						CClassifyingCallbackDataContext *possCallbackContext = new CClassifyingCallbackDataContext(ontClassItem,ccoe->getOntology());
 						possCallback->setCallbackDataContext(possCallbackContext);
diff --git a/Source/Reasoner/Classifier/CSubsumptionClassifierThread.h b/Source/Reasoner/Classifier/CSubsumptionClassifierThread.h
index e914800b..5f07ae54 100644
--- a/Source/Reasoner/Classifier/CSubsumptionClassifierThread.h
+++ b/Source/Reasoner/Classifier/CSubsumptionClassifierThread.h
@@ -130,7 +130,8 @@ namespace Konclude {
 
 					QList<COntologyClassificationItem*> ontItemList;
 					QList<COntologyClassificationItem*> processingOntItemList;
-					QHash<CConcreteOntology*,COntologyClassificationItem*> ontItemHash;
+					QHash<QPair<CConcreteOntology*,cint64>,COntologyClassificationItem*> ontItemHash;
+					QList<COntologyClassificationItem*> mStaleClassItems;
 					cint64 mClassificationCount;
 
 					CReasonerManager* reasoner;
diff --git a/Source/Reasoner/Consistiser/CPrecomputationTestingItem.cpp b/Source/Reasoner/Consistiser/CPrecomputationTestingItem.cpp
index 029bec03..cd9a2544 100644
--- a/Source/Reasoner/Consistiser/CPrecomputationTestingItem.cpp
+++ b/Source/Reasoner/Consistiser/CPrecomputationTestingItem.cpp
@@ -19,6 +19,7 @@
  */
 
 #include "CPrecomputationTestingItem.h"
+#include <atomic>
 
 
 namespace Konclude {
@@ -29,6 +30,8 @@ namespace Konclude {
 
 
 			CPrecomputationTestingItem::CPrecomputationTestingItem(COntologyPrecomputationItem* preCompItem, CConsistenceCalculationStatisticsCollection* statisticCollection) {
+				static std::atomic<cint64> sNextGeneration(0);
+				mGeneration = ++sNextGeneration;
 				mOntologyPreCompItem = preCompItem;
 				mStatisticCollection = statisticCollection;
 			}
diff --git a/Source/Reasoner/Consistiser/CPrecomputationTestingItem.h b/Source/Reasoner/Consistiser/CPrecomputationTestingItem.h
index 99bc5ca9..6e04061f 100644
--- a/Source/Reasoner/Consistiser/CPrecomputationTestingItem.h
+++ b/Source/Reasoner/Consistiser/CPrecomputationTestingItem.h
@@ -22,7 +22,7 @@
 #define KONCLUDE_REASONER_CONSISTISER_CPRECOMPUTATIONTESTINGITEM_H
 
 // Libraries includes
-
+#include <atomic>
 
 // Namespace includes
 #include "ConsistiserSettings.h"
@@ -71,6 +71,8 @@ namespace Konclude {
 					CConsistenceCalculationStatisticsCollection* getUsedStatisticsCollection();
 					CPrecomputationTestingItem* setStatisticCollection(CConsistenceCalculationStatisticsCollection* statisticCollection);
 
+					cint64 mGeneration;
+
 				// protected methods
 				protected:
 
diff --git a/Source/Reasoner/Consistiser/CPrecomputationThread.h b/Source/Reasoner/Consistiser/CPrecomputationThread.h
index 4c4a1166..7f156179 100644
--- a/Source/Reasoner/Consistiser/CPrecomputationThread.h
+++ b/Source/Reasoner/Consistiser/CPrecomputationThread.h
@@ -126,7 +126,7 @@ namespace Konclude {
 					QSet<COntologyPrecomputationItem*> mActiveOntItemSet;
 					QSet<COntologyPrecomputationItem*> mInactiveOntItemSet;
 					QList<COntologyPrecomputationItem*> mProcessingOntItemList;
-					QHash<CConcreteOntology*,COntologyPrecomputationItem*> mOntItemHash;
+					QHash<QPair<CConcreteOntology*,cint64>,COntologyPrecomputationItem*> mOntItemHash;
 					cint64 mPrecomputationCount;
 
 					CReasonerManager* mReasoner;
diff --git a/Source/Reasoner/Ontology/CConcreteOntology.cpp b/Source/Reasoner/Ontology/CConcreteOntology.cpp
index 656c0768..be713438 100644
--- a/Source/Reasoner/Ontology/CConcreteOntology.cpp
+++ b/Source/Reasoner/Ontology/CConcreteOntology.cpp
@@ -121,6 +121,7 @@ namespace Konclude {
 				delete mPreprocessing;
 				delete mRealization;
 				delete mOntologyTriplesData;
+				delete mTax;
 			}
 
 			CConfigurationBase* CConcreteOntology::getConfiguration() {
diff --git a/Source/Reasoner/Ontology/CConcreteOntology.h b/Source/Reasoner/Ontology/CConcreteOntology.h
index c040c86a..a4e5946b 100644
--- a/Source/Reasoner/Ontology/CConcreteOntology.h
+++ b/Source/Reasoner/Ontology/CConcreteOntology.h
@@ -165,6 +165,9 @@ namespace Konclude {
 					COntologyLoadData* getLoadData();
 					CConcreteOntology *setLoadData(COntologyLoadData* loadData);
 
+					cint64 getCallGeneration() const { return mCallGeneration; }
+					void setCallGeneration(cint64 gen) { mCallGeneration = gen; }
+
 
 				// protected methods
 				protected:
@@ -196,6 +199,8 @@ namespace Konclude {
 
 					CONTOLOGYAXIOMSET<CAxiomExpression*>* mAxiomSet;
 
+					cint64 mCallGeneration = 0;
+
 
 				// private methods
 				private:
diff --git a/Source/Reasoner/Preprocess/CPreprocessingThread.h b/Source/Reasoner/Preprocess/CPreprocessingThread.h
index 9703f56a..6272a43d 100644
--- a/Source/Reasoner/Preprocess/CPreprocessingThread.h
+++ b/Source/Reasoner/Preprocess/CPreprocessingThread.h
@@ -118,7 +118,7 @@ namespace Konclude {
 					QSet<COntologyPreprocessingItem*> mActiveOntItemSet;
 					QSet<COntologyPreprocessingItem*> mInactiveOntItemSet;
 					QList<COntologyPreprocessingItem*> mProcessingOntItemList;
-					QHash<CConcreteOntology*,COntologyPreprocessingItem*> mOntItemHash;
+					QHash<QPair<CConcreteOntology*,cint64>,COntologyPreprocessingItem*> mOntItemHash;
 					cint64 mPreprocessingCount;
 
 					CReasonerManager* mReasoner;
diff --git a/Source/Reasoner/Realizer/CRealizerThread.cpp b/Source/Reasoner/Realizer/CRealizerThread.cpp
index 013d96f7..98450852 100644
--- a/Source/Reasoner/Realizer/CRealizerThread.cpp
+++ b/Source/Reasoner/Realizer/CRealizerThread.cpp
@@ -103,7 +103,7 @@ namespace Konclude {
 
 			CRealizerThread* CRealizerThread::processCalculationJob(CSatisfiableCalculationJob* job, COntologyRealizingItem* preCompItem, CRealizingTestingItem* preTestItem) {
 
-				CRealizingCalculatedCallbackEvent* callbackEvent = new CRealizingCalculatedCallbackEvent(this,job,preTestItem);
+				CRealizingCalculatedCallbackEvent* callbackEvent = new CRealizingCalculatedCallbackEvent(this,job,preTestItem,preTestItem->mGeneration);
 				preCompItem->getRealizingTestingItemSet()->insert(preTestItem);
 
 				mCurrRunningTestParallelCount++;
@@ -154,7 +154,7 @@ namespace Konclude {
 				} else if (type == CRealizationMessageEvent::EVENTTYPE) {
 					CRealizationMessageEvent* otcme = (CRealizationMessageEvent *)event;
 					CConcreteOntology* ontology = otcme->getOntology();
-					COntologyRealizingItem* ontRealItem = mOntItemHash.value(ontology,0);
+					COntologyRealizingItem* ontRealItem = mOntItemHash.value({ontology, ontology->getCallGeneration()},0);
 					if (ontRealItem) {
 						CRealizationMessageData* messageData = otcme->getMessageData();
 						CMemoryPool* memoryPools = otcme->getMemoryPool();
@@ -172,10 +172,12 @@ namespace Konclude {
 					QList<COntologyProcessingRequirement*>* requirementList = poe->getRealizingRequirementList();
 
 
-					COntologyRealizingItem* item = mOntItemHash.value(ontology);
+					COntologyRealizingItem* item = mOntItemHash.value({ontology, ontology->getCallGeneration()});
 					if (!item) {						
+						for (auto* staleItem : mInactiveOntItemSet) { delete staleItem; }
+						mInactiveOntItemSet.clear();
 						item = initializeOntologyRealizingItem(ontology,configBase);
-						mOntItemHash.insert(ontology,item);
+						mOntItemHash.insert({ontology, ontology->getCallGeneration()},item);
 						mActiveOntItemSet.insert(item);
 						mProcessingOntItemList.append(item);
 					} else {
@@ -211,7 +213,7 @@ namespace Konclude {
 					CConcreteOntology* ontology = cpoe->getOntology();
 					CCallbackData* realizeedCallbackData = cpoe->getRealizedCallbackData();
 					CCallbackData* installedCallbackData = cpoe->getCallbackPossibleCallbackData();
-					COntologyRealizingItem* item = mOntItemHash.value(ontology);
+					COntologyRealizingItem* item = mOntItemHash.value({ontology, ontology->getCallGeneration()});
 					if (item) {
 						addOntologyRealizingCallback(item,realizeedCallbackData);
 						if (installedCallbackData) {
diff --git a/Source/Reasoner/Realizer/CRealizerThread.h b/Source/Reasoner/Realizer/CRealizerThread.h
index fc09db28..10227216 100644
--- a/Source/Reasoner/Realizer/CRealizerThread.h
+++ b/Source/Reasoner/Realizer/CRealizerThread.h
@@ -123,7 +123,7 @@ namespace Konclude {
 					QSet<COntologyRealizingItem*> mActiveOntItemSet;
 					QSet<COntologyRealizingItem*> mInactiveOntItemSet;
 					QList<COntologyRealizingItem*> mProcessingOntItemList;
-					QHash<CConcreteOntology*,COntologyRealizingItem*> mOntItemHash;
+					QHash<QPair<CConcreteOntology*,cint64>,COntologyRealizingItem*> mOntItemHash;
 					cint64 mRealizingCount;
 
 					CReasonerManager* mReasoner;
diff --git a/Source/Reasoner/Realizer/CRealizingTestingItem.cpp b/Source/Reasoner/Realizer/CRealizingTestingItem.cpp
index ad72cb43..53e54f16 100644
--- a/Source/Reasoner/Realizer/CRealizingTestingItem.cpp
+++ b/Source/Reasoner/Realizer/CRealizingTestingItem.cpp
@@ -19,6 +19,7 @@
  */
 
 #include "CRealizingTestingItem.h"
+#include <atomic>
 
 
 namespace Konclude {
@@ -29,6 +30,8 @@ namespace Konclude {
 
 
 			CRealizingTestingItem::CRealizingTestingItem(COntologyRealizingItem* preproItem, COntologyRealizingDynamicRequirmentProcessingData* procData) {
+				static std::atomic<cint64> sNextGeneration(0);
+				mGeneration = ++sNextGeneration;
 				mOntologyPreproItem = preproItem;
 				if (procData) {
 					mProcDataList.append(procData);
diff --git a/Source/Reasoner/Realizer/CRealizingTestingItem.h b/Source/Reasoner/Realizer/CRealizingTestingItem.h
index 19765d5a..00d712c3 100644
--- a/Source/Reasoner/Realizer/CRealizingTestingItem.h
+++ b/Source/Reasoner/Realizer/CRealizingTestingItem.h
@@ -22,7 +22,7 @@
 #define KONCLUDE_REASONER_REALIZER_CREALIZINGTESTINGITEM_H
 
 // Libraries includes
-
+#include <atomic>
 
 // Namespace includes
 #include "RealizerSettings.h"
@@ -66,6 +66,8 @@ namespace Konclude {
 					QList<COntologyRealizingDynamicRequirmentProcessingData*> getProcessingDataList();
 					CRealizingTestingItem* addProcessingData(COntologyRealizingDynamicRequirmentProcessingData* procData);
 
+					cint64 mGeneration;
+
 				// protected methods
 				protected:
 
diff --git a/Source/Scheduler/CSingleThreadTaskProcessorUnit.cpp b/Source/Scheduler/CSingleThreadTaskProcessorUnit.cpp
index aec293bd..efc88833 100644
--- a/Source/Scheduler/CSingleThreadTaskProcessorUnit.cpp
+++ b/Source/Scheduler/CSingleThreadTaskProcessorUnit.cpp
@@ -266,10 +266,12 @@ namespace Konclude {
 		CThreadActivator* CSingleThreadTaskProcessorUnit::signalizeEvent() {
 			mEventSignalized = true;
 			if (mProcessingBlocked) {
-				// reactivate processing
+				// WASM single-threaded: no background thread to wake via semaphore.
+				// Drive the processing loop directly. Guard with the tag pair so we
+				// don't re-enter when signalizeEvent() is called from within the loop.
 				if (mLastProcessingStartedTag == mLastProcessingStartRequestTag) {
 					++mLastProcessingStartRequestTag;
-					mProcessingWakeUpSemaphore.release();
+					processingLoop();
 				}
 			}
 			return this;
@@ -295,19 +297,13 @@ namespace Konclude {
 			bool eventSafeguardProcessed = false;
 			while (!mProcessingStopped) {
 				if (!mTaskProcessingQueue && mProcessingBlocked) {
-					// block until new events or task are available
-#ifdef KONCLUDE_SCHEDULER_TASK_THREADS_TIME_STATISTICS
-					mStatComputionTime += mComputionTimer.elapsed();
-					mBlockingTimer.start();
-#endif
+					if (!mEventSignalized) {
+						// WASM single-threaded: no blocking possible.  Exit and let
+						// signalizeEvent() restart us when new work arrives.
+						return true;
+					}
+					// Events are pending — fall through as if we "woke up".
 					INCTASKPROCESSINGSTAT(mStats.incStatisticThreadsBlockedCount(1));
-					mThreadBlocked = true;
-					mProcessingWakeUpSemaphore.acquire(1);
-					mThreadBlocked = false;
-#ifdef KONCLUDE_SCHEDULER_TASK_THREADS_TIME_STATISTICS
-					mStatBlockingTime += mBlockingTimer.elapsed();
-					mComputionTimer.start();
-#endif
 					mLastProcessingStartedTag = mLastProcessingStartRequestTag;
 				}
 				mProcessingBlocked = false;
diff --git a/Source/Scheduler/CSingleThreadTaskProcessorUnit.h b/Source/Scheduler/CSingleThreadTaskProcessorUnit.h
index dfdeab30..26d8eda8 100644
--- a/Source/Scheduler/CSingleThreadTaskProcessorUnit.h
+++ b/Source/Scheduler/CSingleThreadTaskProcessorUnit.h
@@ -22,6 +22,7 @@
 #define KONCLUDE_SCHEDULER_CSINGLETHREADTASKPROCESSORUNIT_H
 
 // Libraries includes
+#include <atomic>
 
 
 // Namespace includes
@@ -174,7 +175,7 @@ namespace Konclude {
 
 				bool mEventSignalized;
 				bool mProcessingBlocked;
-				cint64 mLastProcessingStartRequestTag;
-				cint64 mLastProcessingStartedTag;
+				std::atomic<cint64> mLastProcessingStartRequestTag;
+				std::atomic<cint64> mLastProcessingStartedTag;
 				cint64 mTaskProcessingCount;
 				cint64 mTaskSchedulingCount;
diff --git a/Source/Reasoner/Generator/CConcreteOntologyRedlandTriplesDataExpressionMapper.h b/Source/Reasoner/Generator/CConcreteOntologyRedlandTriplesDataExpressionMapper.h
index 8df9ce6d..a5293758 100644
--- a/Source/Reasoner/Generator/CConcreteOntologyRedlandTriplesDataExpressionMapper.h
+++ b/Source/Reasoner/Generator/CConcreteOntologyRedlandTriplesDataExpressionMapper.h
@@ -79,6 +78,7 @@ namespace Konclude {
 
 					virtual bool mapTriples(CConcreteOntology* updateConcreteOntology, COntologyTriplesData* ontologyTripleData);
 
+					void setConfExtractSimpleABoxAssertions(bool v) { mConfExtractSimpleABoxAssertions = v; }
 
 
 

